src/utils/database-common/src/transactions/db_transaction_manager.rs

```diff
@@ -34,49 +34,74 @@ pub trait DatabaseTransactionManager: Send + Sync {
-    match result {
-        // In case everything succeeded, commit the transaction
-        Ok(_) => {
-            db_transaction_manager
-                .commit_transaction(transaction_ref)
-                .await?;
-            Ok(())
-        }
+pub struct DatabaseTransactionRunner<HFutResultE = InternalError> {
+    _phantom: PhantomData<HFutResultE>,
+}
 
-        // Otherwise, do an explicit rollback
-        Err(e) => {
-            db_transaction_manager
-                .rollback_transaction(transaction_ref)
-                .await?;
-            Err(e)
+impl<HFutResultE> DatabaseTransactionRunner<HFutResultE> {
+    pub async fn run_transactional<H, HFut, HFutResultT>(
+        base_catalog: &Catalog,
+        callback: H,
+    ) -> Result<HFutResultT, HFutResultE>
+    where
+        H: FnOnce(Catalog) -> HFut + Send,
+        HFut: std::future::Future<Output = Result<HFutResultT, HFutResultE>> + Send,
+        HFutResultE: From<InternalError>,
+    {
+        // Extract transaction manager, specific for the database
+        let db_transaction_manager = base_catalog
+            .get_one::<dyn DatabaseTransactionManager>()
+            .unwrap();
+
+        // Start transaction
+        let transaction_ref = db_transaction_manager.make_transaction_ref().await?;
+
+        // Create a chained catalog for transaction-aware components,
+        // but keep a local copy of a transaction pointer
+        let chained_catalog = CatalogBuilder::new_chained(base_catalog)
+            .add_value(transaction_ref.clone())
+            .build();
+
+        // Run transactional code in the callback
+        let result = callback(chained_catalog).await;
+
+        // Commit or rollback transaction depending on the result
+        match result {
+            // In case everything succeeded, commit the transaction
+            Ok(res) => {
+                db_transaction_manager
+                    .commit_transaction(transaction_ref)
+                    .await?;
+                Ok(res)
+            }
+
+            // Otherwise, do an explicit rollback
+            Err(e) => {
+                db_transaction_manager
+                    .rollback_transaction(transaction_ref)
+                    .await?;
+                Err(e)
+            }
         }
     }
+
+    pub async fn run_transactional_with<Store, H, HFut, HFutResultT>(
+        base_catalog: &Catalog,
+        callback: H,
+    ) -> Result<HFutResultT, HFutResultE>
+    where
+        Store: 'static + ?Sized + Send + Sync,
+        H: FnOnce(Arc<Store>) -> HFut + Send,
+        HFut: std::future::Future<Output = Result<HFutResultT, HFutResultE>> + Send,
+        HFutResultE: From<InternalError>,
+    {
+        Self::run_transactional(base_catalog, |updated_catalog| async move {
+            let event_store = updated_catalog.get_one::<Store>().int_err()?;
+
+            callback(event_store).await
+        })
+        .await
+    }
```

Comment:
The original idea was to give the callback the entire catalog, as it might need a lot more than just 1 repository interface in a more complex scenario. This mightly be slightly simplifying a specific type of case.